### ISSUE TITLE
Partial CHaP release

### ISSUE THREAD

**Issue by locallycompact** [cardano-foundation/cardano-wallet#4411]
### The problem that you wish to solve

Could we get a release of the cardano-wallet cli on CHaP?

### Description

As above,

## Acceptance criteria

- [ ] `cardano-wallet` executable is published on CHaP (https://chap.intersectmbo.org/) and resolvable as a dependency via `source-repository-package` or direct Hackage/CHaP pin
- [ ] A release note / PR description links to the CHaP entry for the published release
- [ ] CI includes a check or documented command that installs the wallet CLI from CHaP and runs `cardano-wallet --version`

**Comment by HeinrichApfelmus** [comment]
May I inquire about the purpose? The main reason for releasing the `cardano-wallet` executable in CHaP would be to be able to depend upon it from other `.cabal` packages — do you have a particular cabal package in mind where you want to do this?

**Comment by locallycompact** [comment]
To include in horizon so it can be installed on a system without use of IFD, using the same build plan and dependency footprint as all the other cli tools.

https://gitlab.horizon-haskell.net/package-sets/horizon-cardano
https://gitlab.horizon-haskell.net/package-sets/horizon-cardano/-/blob/master/horizon.dhall

**Comment by euonymos** [comment]
Some tools, let's say [Atlas PAB](https://github.com/geniusyield/atlas) use `cardano-wallet` as a Haskell library, and they are forced to copy its gory `cabal.project` file which sucks. Their dependants have to do the same. Does is sound like a good argument to you @HeinrichApfelmus?

**Comment by abailly** [comment]
Do we need release to CHaP, which will be somewhat complicated because CHaP is controlled by another organisation, or would a release to https://hackage.haskell.org be good?

**Comment by abailly** [comment]
@euonymos Also, I would be interested in understanding which part of the cardano-wallet API you are using. A quick search in the PAB repository only showed `cardano-wallet-primitives` but I may have overlooked other uses.

**Comment by locallycompact** [comment]
@abailly It would be unsatisfying to release to hackage in my opinion because the dependencies would not resolve without CHaP. Maybe cardano-foundation deserves its own release index?

**Comment by sourabhxyz** [comment]
@abailly Thank you for your interest in this issue. I would like to chime in regarding some of your remarks.

> @euonymos Also, I would be interested in understanding which part of the cardano-wallet API you are using. A quick search in the PAB repository only showed cardano-wallet-primitives but I may have overlooked other uses.

@euonymos remark was wrt [Atlas](https://github.com/geniusyield/atlas), an Haskell based transaction building tool. Inside Atlas, we use `cardano-coin-selection` package from `cardano-wallet`. Due to this dependency we have to import `cabal.project` details of `cardano-wallet` inside Atlas, [here](https://github.com/geniusyield/atlas/blob/233122f41304647c194681cbf26c36405440a51d/cabal.project#L95-L194). Likewise, any project using Atlas in their stack have to do the same which is a huge bloat atm.

> Do we need release to CHaP, which will be somewhat complicated because CHaP is controlled by another organisation, or would a release to https://hackage.haskell.org/ be good?

Isn't `cardano-wallet` itself dependent upon chap? I mean, how could then this package be uploaded onto Hackage as it would not build in the standard Hackage environment since Hackage has no access to chap? I could be missing something here.

**Comment by abailly** [comment]
@locallycompact @sourabhxyz I agree depending on CHaP from Hackage is a major problem, I will raise an issue in the relevant repository to understand how we can streamline the release process there from here.

**Comment by abailly** [comment]
So there's a script for adding packages to CHaP that we could use: https://github.com/IntersectMBO/cardano-haskell-packages?tab=readme-ov-file#-from-github

**Comment by abailly** [comment]
I did a quick analysis of c-w dependency graph using `cabal-plan` and here is the local dependency graph:

![deps](https://github.com/user-attachments/assets/86bb3c55-4f32-4d53-9d92-deb80e89afa9)

Restricting to the `cardano-coin-selection` package, this yields:

![deps](https://github.com/user-attachments/assets/789d20f0-9fb1-45aa-8318-070422d1785e)

@sourabhxyz Would that help if we published only the latter set? This seems a low-hanging fruit we can solve relatively easily.

@locallycompact Did you have request from people using horizon to be able to pull c-w executable from CHaP?

**Comment by locallycompact** [comment]
It was originally a request from myself to myself but I suspect having atlas in horizon would also find use.

**Comment by sourabhxyz** [comment]
@abailly

> @sourabhxyz Would that help if we published only the latter set? This seems a low-hanging fruit we can solve relatively easily.

Yes indeed, that's a great idea!

**Comment by abailly** [comment]
> Yes indeed, that's a great idea!

Would the latest version (2024.11.18) be good or do you need an earlier version in order to take advantage of this?

**Comment by sourabhxyz** [comment]
> Would the latest version (2024.11.18) be good or do you need an earlier version in order to take advantage of this?

I'll need some time to evaluate it, but likely we can manage with latest version.

**Comment by euonymos** [comment]
Thank you a lot @abailly, I will try to find time to wire in chap's libs.

**Comment by abailly** [comment]
So it seems this might be more complicated than expected because of the way the project is structured: The bounds are defined in a `cabal.project` file at the top-level of the project and _not_ in the individual `.cabal` files, but when the individual packages are built against CHaP repository, dependency resolution might lead to pulling in versions which are not supported because of the lack of bounds at the package description level.

We probably need to improve the structure of the project, including removing source-repository dependencies, before considering CHaP publication IIUC the problem.

cc @hamishmack @HeinrichApfelmus @paolino

**Comment by HeinrichApfelmus** [comment]
Working on publishing `cardano-wallet-read` to CHaP: https://github.com/IntersectMBO/cardano-haskell-packages/pull/1003

**Comment by HeinrichApfelmus** [comment]
I have published `cardano-wallet-read-1.0.0.0` to CHaP by merging https://github.com/IntersectMBO/cardano-haskell-packages/pull/949 . ✅

**Comment by abailly** [comment]
To generate dependency graph for local package
```
cabal-plan dot --hide-global --hide-builtin | tred | gvpr -f subgraph.gvpr -a "cardano-coin-selection-0.2025.1.9" | dot -Tpdf -odeps.pdf
```
with  `subgraph.gvpr` being
```
/* Construct subgraph reachable from node ARGV[0] by forward edges */
BEG_G {
  node_t r = node($,ARGV[0]);

  $tvroot = r;
  $tvtype = TV_fwd;
  //print ("//  Staring node: ", r.name);
}
N{
  print ("// subgraph node: ", $.name);
  $tvroot=NULL;
  subnode($T,$);
}
E{
  print ("// subgraph edge: ", $.name);
  subedge($T,$);
}

```

**Comment by paolino** [comment]
**Audit note (2026-02-12):** Prerequisite #4958 (set dependency bounds) is partially done — some packages have proper bounds, others still have unbounded deps. Until all packages pass `cabal check` with proper bounds, CHaP publishing will remain blocked.

---

🤖 Claude opus 5 agent

**Comment by paolino** [comment]
Triaging open issues: this predates the current cardano-wallet release cadence. Is "partial CHaP release" still a goal, or has the workflow since evolved to full CHaP publishing on release? Close if no longer in scope.

---

🤖 Claude opus 5 agent

**Comment by paolino** [comment]
Triage against `master` @ `eeffc429e7`:

Still open — nothing on master publishes `cardano-wallet` to CHaP; the repo only *consumes* CHaP.

- `cabal.project:40-41` declares `repository cardano-haskell-packages` / `url: https://chap.intersectmbo.org/` — a dependency *source*, not a publish target.
- Release workflows (`.github/workflows/release.yml`, `publish-release.yml`, `publish.yml`) cover GitHub releases, Docker tags and docs only; `git grep -i chap origin/master -- .github/workflows/ scripts/release/` finds no publish-to-CHaP step (sole hit is a `ci.yml:708` comment about the CHaP index for HLS).
- No CI check installs the wallet CLI from CHaP, so the acceptance criteria are unmet.

— qwen
